src/main/java/hudson/maven/reporters/TestMojo.java

```diff
@@ -167,21 +168,34 @@ public boolean canRunTests(MojoInfo mojo) {
     }
     
     @CheckForNull public Iterable<File> getReportFiles(MavenProject pom, MojoInfo mojo) throws ComponentConfigurationException {
+        File reportsDir = getReportsDirectory(pom, mojo);
+        if (reportsDir != null && reportsDir.exists())
+            return getReportFiles(reportsDir, getFileSet(reportsDir));
+        
+        return null;
+    }
+
+    /**
+     * Returns the location of test reports created by the specified MOJO.
+     * @param pom The project model.
+     * @param mojo The MOJO.
+     * @return The directory containing the test reports.
+     * @throws ComponentConfigurationException if unable to retrieve the report directory from the MOJO configuration.
+     */
+    protected File getReportsDirectory(MavenProject pom, MojoInfo mojo) throws ComponentConfigurationException {
+    	// [JENKINS-31258] Allow unknown MOJOs to contribute test results in arbitrary locations by setting a Maven property.
+		String reportsDirectoryOverride = getReportsDirectoryOverride(mojo);
+		if (reportsDirectoryOverride != null)
+			return mojo.expressionEvaluator.alignToBaseDirectory(new File(reportsDirectoryOverride));
+
         if (this.reportDirectoryConfigKey != null) {
             File reportsDir = mojo.getConfigurationValue(this.reportDirectoryConfigKey, File.class);
-            if (reportsDir != null && reportsDir.exists()) {
-                return getReportFiles(reportsDir, getFileSet(reportsDir));
-            } 
-            
+            if (reportsDir != null)
+                return reportsDir;
```

Comment:
This plugin is a splitt-off of the core, hence it follows the Oracle codestyle. In this style brackets are required for single-liners